concurrency/WFLY-12896_EE_Concurrency_Hung_Tasks_Termination.adoc

```diff
@@ -0,0 +1,95 @@
+= EE Concurrency Hung Tasks Termination
+:author:            Eduardo Martins
+:email:             emartins@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+The managed executors in the EE subsystem are capable of identifying tasks which are running for a long unexpected time, also know as hung tasks, and this proposal specifies new functionality that will allows users to manually and automatically terminate such tasks.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12896[WFLY-12896]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1397[EAP7-1397]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[x] Engineering
+
+[ ] QE
+
+=== Affected Projects or Components
+
+* WildFly
+
+=== Other Interested Projects
+
+* NONE
+
+== Requirements
+
+=== Hard Requirements
+
+* Add the following attribute to the https://wildscribe.github.io/WildFly/18.0/subsystem/ee/managed-executor-service/[/subsystem=ee/managed-executor-service resource]:
+** *hung-task-termination-period*: the period, in milliseconds, for hung tasks automatic termination.
+
+CLI to read the attribute from the Managed Executor Service resource named *default*:
+----
+/subsystem=ee/managed-executor-service=default:read-attribute(name=hung-task-termination-period)
+----
+
+* Add the following operation to the https://wildscribe.github.io/WildFly/18.0/subsystem/ee/managed-executor-service/[/subsystem=ee/managed-executor-service resource]:
+** *terminate-hung-tasks*: manually terminates all executor tasks that are hung.
+
+CLI to execute the operation from the Managed Executor Service resource named *default*:
+----
+/subsystem=ee/managed-executor-service=default:terminate-hung-tasks()
```

Comment:
IMHO I don't see much value in leaving tasks that are hung alive so I would prefer to keep the simplicity of the feature, having a match of what the new management op and the automatic periodic process does (terminate all hung tasks).

Reply:
@emmartins actually I was asking because of a scenario I experienced myself; let me depict it: we had something like 10 long running tasks marked as "hung" and we knew they were hung because they generated high concurrency in the the database connected to the application server; this pile of tasks was generated by manual retries; we wanted to get rid of such concurrency killing 9 of them to enable the one left to complete it's work: it would have been too expensive to restart it; our criteria back then was to kill the most recent 9, but that was just a simplistic criteria we used because we lacked further info on the tasks; the other solution would have been to engage a db administrator to work at the db level and we had none available back then; the vendor was Weblogic, and if remember correctly we could kill hung tasks selectively; anyway, if you think this is an over specific scenario I won't insist further

Reply:
The managed tasks can have a listener attached with logic to deal with termination on their own, we also have a runtime stat providing the number of hung tasks, which would need to be reworked since having the stat and a retrieve tasks attr/op would be confusing IMHO. Let's leave this extended functionality for later, if customers ask for it.

Reply:
agree